File: classifications/old_indoor_test.go
package classifications_test

import (
	"testing"

	"github.com/jatkinson1000/archeryutils/classifications"
	"github.com/jatkinson1000/archeryutils/rounds"
	"github.com/jatkinson1000/archeryutils/targets"
)

func TestOldIndoorClassificationScoresAges(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	want := []int{592, 582, 554, 505, 432, 315, 195, 139}
	// All ages map to Adult for old indoor
	got, err := classifications.OldIndoorClassificationScores(r, classifications.Recurve, classifications.Male, classifications.Adult, true)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertIntSliceEqual(t, got, want)
}

func TestOldIndoorClassificationScoresGender(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	want := []int{582, 569, 534, 479, 380, 255, 139, 93}
	got, err := classifications.OldIndoorClassificationScores(r, classifications.Recurve, classifications.Female, classifications.Adult, true)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertIntSliceEqual(t, got, want)
}

func TestOldIndoorClassificationScoresBowstyles(t *testing.T) {
	tests := []struct {
		bowstyle classifications.Bowstyle
		gender   classifications.Gender
		want     []int
	}{
		{classifications.Compound, classifications.Male, []int{581, 570, 554, 529, 484, 396, 279, 206}},
		{classifications.Compound, classifications.Female, []int{570, 562, 544, 509, 449, 347, 206, 160}},
	}
	for _, tc := range tests {
		t.Run(tc.bowstyle.String()+"_"+tc.gender.String(), func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.OldIndoorClassificationScores(r, tc.bowstyle, tc.gender, classifications.Adult, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestOldIndoorClassificationScoresTripleFaces(t *testing.T) {
	// Triple and spot variants: reversed expected (ascending → descending)
	tests := []struct {
		round string
		want  []int // descending order (A first)
	}{
		{"portsmouth_triple", []int{581, 570, 554, 529, 484, 396, 279, 206}},
		{"worcester_5_centre", []int{300, 299, 289, 264, 226, 162, 96, 65}},
		{"wa18_triple", []int{568, 558, 537, 486, 370, 203, 100, 63}},
		{"wa18", []int{568, 558, 537, 493, 420, 295, 173, 117}},
	}
	for _, tc := range tests {
		t.Run(tc.round, func(t *testing.T) {
			r := indoorRound(t, tc.round)
			got, err := classifications.OldIndoorClassificationScores(r, classifications.Compound, classifications.Male, classifications.Adult, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestOldIndoorClassificationScoresInvalidBowstyle(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	_, err := classifications.OldIndoorClassificationScores(r, classifications.Barebow, classifications.Male, classifications.Adult, true)
	if err == nil {
		t.Fatal("expected error for invalid (Barebow) bowstyle")
	}
}

func TestOldIndoorClassificationScoresInvalidAge(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	_, err := classifications.OldIndoorClassificationScores(r, classifications.Recurve, classifications.Male, classifications.Under12, true)
	if err == nil {
		t.Fatal("expected error for invalid age (Under12)")
	}
}

func TestOldIndoorClassificationScoresInvalidRound(t *testing.T) {
	r := outdoorRound(t, "wa1440_90")
	_, err := classifications.OldIndoorClassificationScores(r, classifications.Recurve, classifications.Male, classifications.Adult, true)
	if err == nil {
		t.Fatal("expected error for outdoor round in old indoor")
	}
}

func TestOldIndoorClassificationScoresNonStrict(t *testing.T) {
	// Frostbite: 36 arrows at 30m 80cm outdoor — not an indoor round
	frostbite := mustOutdoorPass(t, 36, "10_zone", targets.CM(80), targets.Metres(30))
	r, err := rounds.NewRound("Frostbite", []*rounds.Pass{frostbite})
	if err != nil {
		t.Fatalf("NewRound: %v", err)
	}
	want := []int{357, 351, 336, 310, 269, 195, 110, 68}
	got, err := classifications.OldIndoorClassificationScores(r, classifications.Compound, classifications.Male, classifications.Adult, false)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertIntSliceEqual(t, got, want)
}

func TestCalculateOldIndoorClassification(t *testing.T) {
	tests := []struct {
		score  float64
		gender classifications.Gender
		want   string
	}{
		{400, classifications.Male, "F"},
		{337, classifications.Female, "F"},
		{592, classifications.Male, "A"},
		{582, classifications.Female, "A"},
		{581, classifications.Male, "C"},
		{120, classifications.Male, "UC"},
		{1, classifications.Male, "UC"},
	}
	for _, tc := range tests {
		t.Run(tc.gender.String()+"_"+tc.want, func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.CalculateOldIndoorClassification(tc.score, r, classifications.Recurve, tc.gender, classifications.Adult, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got != tc.want {
				t.Errorf("score=%.0f: got %q, want %q", tc.score, got, tc.want)
			}
		})
	}
}

func TestCalculateOldIndoorClassificationInvalidScore(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	for _, score := range []float64{1000, 601, -1, -100} {
		_, err := classifications.CalculateOldIndoorClassification(score, r, classifications.Recurve, classifications.Male, classifications.Adult, true)
		if err == nil {
			t.Errorf("expected error for score=%.0f", score)
		}
	}
}

func TestCoaxOldIndoorGroup(t *testing.T) {
	tests := []struct {
		bowstyle     classifications.Bowstyle
		wantBowstyle classifications.Bowstyle
		wantAge      classifications.Age
	}{
		{classifications.Compound, classifications.Compound, classifications.Adult},
		{classifications.CompoundLimited, classifications.Compound, classifications.Adult},
		{classifications.CompoundBarebow, classifications.Compound, classifications.Adult},
		{classifications.Recurve, classifications.Recurve, classifications.Adult},
		{classifications.Barebow, classifications.Recurve, classifications.Adult},
		{classifications.Longbow, classifications.Recurve, classifications.Adult},
	}
	for _, tc := range tests {
		t.Run(tc.bowstyle.String(), func(t *testing.T) {
			cat := classifications.CoaxOldIndoorGroup(tc.bowstyle, classifications.Male, classifications.Under18)
			if cat.Bowstyle != tc.wantBowstyle {
				t.Errorf("bowstyle: got %s, want %s", cat.Bowstyle, tc.wantBowstyle)
			}
			if cat.AgeGroup != tc.wantAge {
				t.Errorf("age: got %s, want %s", cat.AgeGroup, tc.wantAge)
			}
		})
	}
}

File: ts/src/classifications/old-indoor.test.ts
import { describe, it, expect } from 'vitest';
import {
  oldIndoorClassificationScores,
  calculateOldIndoorClassification,
  coaxOldIndoorGroup,
} from './old-indoor.js';
import { Gender, Age, Bowstyle } from './data.js';
import { agbIndoorRounds, waIndoorRounds, type Round, atTarget } from '../rounds.js';
import { ScoringSystem, cm, metres } from '../targets.js';

// ---- helpers ----

function indoorRound(name: string): Round {
  return (agbIndoorRounds().get(name) ?? waIndoorRounds().get(name))!;
}

// ---- ages (all map to Adult for old indoor) ----

describe('oldIndoorClassificationScores ages all → Adult', () => {
  it('Adult recurve male portsmouth', () => {
    const want = [592, 582, 554, 505, 432, 315, 195, 139];
    const scores = oldIndoorClassificationScores(
      indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Male, Age.Adult, true,
    );
    expect(scores).toEqual(want);
  });
});

// ---- gender ----

describe('oldIndoorClassificationScores gender', () => {
  it('female recurve portsmouth', () => {
    const want = [582, 569, 534, 479, 380, 255, 139, 93];
    const scores = oldIndoorClassificationScores(
      indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Female, Age.Adult, true,
    );
    expect(scores).toEqual(want);
  });
});

// ---- bowstyles ----

describe('oldIndoorClassificationScores bowstyles', () => {
  const cases: [Bowstyle, Gender, number[]][] = [
    [Bowstyle.Compound, Gender.Male,   [581, 570, 554, 529, 484, 396, 279, 206]],
    [Bowstyle.Compound, Gender.Female, [570, 562, 544, 509, 449, 347, 206, 160]],
  ];
  for (const [bowstyle, gender, want] of cases) {
    it(`${bowstyle} ${gender}`, () => {
      const scores = oldIndoorClassificationScores(
        indoorRound('portsmouth'), bowstyle, gender, Age.Adult, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- triple faces / spot variants ----

describe('oldIndoorClassificationScores triple faces', () => {
  const cases: [string, number[]][] = [
    ['portsmouth_triple', [581, 570, 554, 529, 484, 396, 279, 206]],
    ['worcester_5_centre', [300, 299, 289, 264, 226, 162, 96, 65]],
    ['wa18_triple', [568, 558, 537, 486, 370, 203, 100, 63]],
    ['wa18',        [568, 558, 537, 493, 420, 295, 173, 117]],
  ];
  for (const [roundName, want] of cases) {
    it(roundName, () => {
      const scores = oldIndoorClassificationScores(
        indoorRound(roundName), Bowstyle.Compound, Gender.Male, Age.Adult, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- non-strict ----

describe('oldIndoorClassificationScores non-strict', () => {
  it('frostbite (outdoor) with non-strict', () => {
    const frostbite: Round = {
      name: 'Frostbite',
      codename: 'frostbite',
      passes: [atTarget(36, ScoringSystem.TenZone, cm(80), metres(30), false)],
      nArrows: 36,
    };
    const want = [357, 351, 336, 310, 269, 195, 110, 68];
    const scores = oldIndoorClassificationScores(
      frostbite, Bowstyle.Compound, Gender.Male, Age.Adult, false,
    );
    expect(scores).toEqual(want);
  });
});

// ---- errors ----

describe('oldIndoorClassificationScores errors', () => {
  it('throws for invalid bowstyle (Barebow)', () => {
    expect(() => oldIndoorClassificationScores(
      indoorRound('portsmouth'), Bowstyle.Barebow, Gender.Male, Age.Adult, true,
    )).toThrow();
  });
  it('throws for invalid age (Under12)', () => {
    expect(() => oldIndoorClassificationScores(
      indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Male, Age.Under12, true,
    )).toThrow();
  });
  it('throws for outdoor round', () => {
    // use an outdoor round codename not in indoor
    const outdoorRound: Round = {
      name: 'WA1440', codename: 'wa1440_90',
      passes: [atTarget(36, ScoringSystem.TenZone, cm(122), metres(90), false)],
      nArrows: 36,
    };
    expect(() => oldIndoorClassificationScores(
      outdoorRound, Bowstyle.Recurve, Gender.Male, Age.Adult, true,
    )).toThrow();
  });
});

// ---- calculateOldIndoorClassification ----

describe('calculateOldIndoorClassification', () => {
  const cases: [number, Gender, string][] = [
    [400, Gender.Male,   'F'],
    [337, Gender.Female, 'F'],
    [592, Gender.Male,   'A'],
    [582, Gender.Female, 'A'],
    [581, Gender.Male,   'C'],
    [120, Gender.Male,   'UC'],
    [  1, Gender.Male,   'UC'],
  ];
  for (const [score, gender, want] of cases) {
    it(`score=${score} ${gender} → ${want}`, () => {
      const cls = calculateOldIndoorClassification(
        score, indoorRound('portsmouth'), Bowstyle.Recurve, gender, Age.Adult, true,
      );
      expect(cls).toBe(want);
    });
  }
});

describe('calculateOldIndoorClassification invalid scores', () => {
  for (const score of [1000, 601, -1, -100]) {
    it(`score=${score} throws`, () => {
      expect(() => calculateOldIndoorClassification(
        score, indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Male, Age.Adult, true,
      )).toThrow();
    });
  }
});

// ---- coaxOldIndoorGroup ----

describe('coaxOldIndoorGroup', () => {
  const cases: [Bowstyle, Bowstyle, Age][] = [
    [Bowstyle.Compound,        Bowstyle.Compound, Age.Adult],
    [Bowstyle.CompoundLimited, Bowstyle.Compound, Age.Adult],
    [Bowstyle.CompoundBarebow, Bowstyle.Compound, Age.Adult],
    [Bowstyle.Recurve,         Bowstyle.Recurve,  Age.Adult],
    [Bowstyle.Barebow,         Bowstyle.Recurve,  Age.Adult],
    [Bowstyle.Longbow,         Bowstyle.Recurve,  Age.Adult],
  ];
  for (const [input, wantBowstyle, wantAge] of cases) {
    it(`${input} → ${wantBowstyle}`, () => {
      const cat = coaxOldIndoorGroup(input, Gender.Male, Age.Under18);
      expect(cat.bowstyle).toBe(wantBowstyle);
      expect(cat.ageGroup).toBe(wantAge);
    });
  }
});
